### PR TITLE
FOR DISCUSSION: Merge ViewportController into DeckGL

### PR DESCRIPTION
**Pull request by 1chandu**
* Move `TransitionManger` and `MapControllerJS` into `DeckGLJS`.
* Add additional prop types and default props to `DeckGL` component.
* Modify viewport transition examples to use new DeckGL props.

- Now the transitions are exposed using DeckGL API instead of experimental `ViewportController`.
- This aligns with future Controller/Controls refactor where `ControllerType` is passed to `DeckGL` as a prop.

Tested with viewport transition and other examples.

### CODE REVIEW

File: src/core/lib/transition-manager.js [hunk 1]
```diff
@@ -11,6 +12,28 @@ export const TRANSITION_EVENTS = {
   IGNORE: 3
 };
 
+const PROP_TYPES = {
+  // transition duration for viewport change
+  transitionDuration: PropTypes.number,
+  // an instance of ViewportTransitionInterpolator, can be used to perform custom transitions.
+  transitionInterpolator: PropTypes.object,
+  // type of interruption of current transition on update.
+  transitionInterruption: PropTypes.number,
```
Comment (ibgreen) [inline src/core/lib/transition-manager.js]: Thought: I suppose these could be strings....

File: src/core/lib/transition-manager.js [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import LinearInterpolator from '../transitions/linear-interpolator';
 import {extractViewportFrom} from '../transitions/transition-utils';
 import assert from 'assert';
+import PropTypes from 'prop-types';
```
Comment (ibgreen) [inline src/core/lib/transition-manager.js]: Nit: Using react's prop-types (even though they are broken our) in non-react apps is a little confusing. Especially as we are drawing a line between the react and non-react parts. Fine for now but something to consider.

File: src/core/lib/transition-manager.js [hunk 3]
```diff
@@ -11,6 +12,28 @@ export const TRANSITION_EVENTS = {
...
+
+  /**
+   * `onViewportChange` callback is fired when the user interacted with the
+   * map. The object passed to the callback contains viewport properties
+   * such as `longitude`, `latitude`, `zoom` etc.
+   */
+  onViewportChange: PropTypes.func
+};
+
 const DEFAULT_PROPS = {
```
Comment (ibgreen) [inline src/core/lib/transition-manager.js]: Thought: Given the extensive props-based API, should this class have a `setProps` method? Many of the other JS classes do. Let's revisit later when we finalize the pure JS API.

File: src/core/pure-js/deck-js.js [hunk 4]
```diff
@@ -163,6 +168,17 @@ export default class DeckGLJS {
...
+
+    if (this._conroller) {
+      this._controller.finalize();
+      this._controller = null;
+    }
+  }
+
+  // Trigger transition, retur true if a new transition is triggered, false otherwise.
+  triggerViewportTransition(nextProps) {
+    return this._transitionManger ?
```
Comment (ibgreen) [inline src/core/pure-js/deck-js.js]: Typo: `Manger`

Nit: `transitionManager.setProps`?

File: src/core/pure-js/deck-js.js [hunk 5]
```diff
@@ -163,6 +168,17 @@ export default class DeckGLJS {
...
     }
+
+    if (this._conroller) {
+      this._controller.finalize();
+      this._controller = null;
+    }
+  }
+
+  // Trigger transition, retur true if a new transition is triggered, false otherwise.
+  triggerViewportTransition(nextProps) {
```
Comment (ibgreen) [inline src/core/pure-js/deck-js.js]: Naming? This sound like an action, not a predicate: how about `isViewportTransitionTriggered`?

File: src/react/deckgl.js [hunk 6]
```diff
@@ -41,6 +41,12 @@ export default class DeckGL extends React.Component {
     }
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    const transitionTriggered = this.deck.triggerViewportTransition(nextProps);
```
Comment (ibgreen) [inline src/react/deckgl.js]: There are many props that can change here that are not related to transitions. 

Coded this way there seems to be a risk that we decline some non-transition related updates that we should be doing.

Think about how we can make the logic clearer here? Or add more comments explaining why this is not an issue?

File: src/react/viewport-controller.js [hunk 7]
```diff
@@ -7,40 +7,43 @@ const {ViewportControls, TransitionManager} = experimental;
 
 import CURSOR from './utils/cursors';
 
-const propTypes = {
+const TRANSITION_PROP_TYPES = {
```
Comment (ibgreen) [inline src/react/viewport-controller.js]: Q: Why is this component being changed? Is it not being deleted?

Comment (1chandu) [reply src/react/viewport-controller.js]: Un-intentional change, and yes, this should be deleted, but before it can be deleted, I also need to update the `layer-browser` example which is using it, will do it in next update to this PR.

**Comment by ibgreen** [comment]
@1chandu - did we decide to close this PR? Since we kept ViewportController as experimental?

**Comment by 1chandu** [comment]
@ibgreen , yes, closing.
